src/app/icd/ICDCheckInSender.cpp

```diff
@@ -0,0 +1,152 @@
+{
+    mResolveInProgress = false;
+
+    ICDMonitoringEntry entry(mSymmetricKeystore);
+
+    const FabricInfo * f = mFabricTable->FindFabricWithCompressedId(peerId.GetCompressedFabricId());
+    ICDMonitoringTable table(*mStorage, f->GetFabricIndex(), CHIP_CONFIG_ICD_CLIENTS_SUPPORTED_PER_FABRIC, mSymmetricKeystore);
+    table.Find(peerId.GetNodeId(), entry);
+
+    if (CHIP_NO_ERROR != SendCheckInMsg(entry, result.address))
+    {
+        ChipLogError(AppServer, "Failed to send the ICD Check-In message");
+    }
+}
+
+void ICDCheckInSender::OnNodeAddressResolutionFailed(const PeerId & peerId, CHIP_ERROR reason)
+{
+    mResolveInProgress = false;
+    ChipLogProgress(AppServer, "Node Address resolution failed for ICD Check-In with Node ID " ChipLogFormatX64,
+                    ChipLogValueX64(peerId.GetNodeId()));
+}
+
+CHIP_ERROR ICDCheckInSender::SendCheckInMsg(ICDMonitoringEntry & entry, const Transport::PeerAddress & addr)
+{
+    uint8_t b[CheckinMessage::sMinPayloadSize] = { 0 };
+    MutableByteSpan output{ b };
+    CHIP_ERROR err;
+
+    // TODO retrieve Check-in counter
+    CounterType counter = 0;
+
+    // Prepare Check-in payload
+    err = CheckinMessage::GenerateCheckinMessagePayload(entry.key, counter, ByteSpan(), output);
+    if (err != CHIP_NO_ERROR)
+    {
+        return err;
+    }
+
+    if (InteractionModelEngine::GetInstance() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    if (InteractionModelEngine::GetInstance()->GetExchangeManager() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    if (InteractionModelEngine::GetInstance()->GetExchangeManager()->GetSessionManager() == nullptr)
+    {
+        return CHIP_ERROR_INTERNAL;
+    }
+
+    Messaging::ExchangeContext * exchangeContext = InteractionModelEngine::GetInstance()->GetExchangeManager()->NewContext(
+        InteractionModelEngine::GetInstance()
+            ->GetExchangeManager()
+            ->GetSessionManager()
+            ->CreateUnauthenticatedSession(addr, GetLocalMRPConfig().ValueOr(GetDefaultMRPConfig()))
+            .Value(),
+        nullptr);
+
+    VerifyOrReturnError(exchangeContext != nullptr, CHIP_ERROR_NO_MEMORY);
+
+    System::PacketBufferHandle buffer = MessagePacketBuffer::NewWithData(output.data(), output.size());
+
+    return exchangeContext->SendMessage(MsgType::ICD_CheckIn, std::move(buffer), Messaging::SendMessageFlags::kNoAutoRequestAck);
+}
+
+CHIP_ERROR ICDCheckInSender::RequestResolve(FabricIndex fabricIndex, NodeId checkInNodeID)
+{
+    VerifyOrReturnError(mFabricTable != nullptr, CHIP_ERROR_INTERNAL);
+    const FabricInfo * fabricInfo = mFabricTable->FindFabricWithIndex(fabricIndex);
+    PeerId peerId(fabricInfo->GetCompressedFabricId(), checkInNodeID);
+
+    AddressResolve::NodeLookupRequest request(peerId);
+    mCurrentNodeId = checkInNodeID;
+
+    CHIP_ERROR err = AddressResolve::Resolver::Instance().LookupNode(request, mAddressLookupHandle);
```

Comment:
What guarantees that the ICD stays in active mode to listen for those lookup results?  I bet nothing does.  Followup needed to make this actually work and make sense; I doubt you want to try to tackle that in this PR.

Reply:
That's an excellent point, and you're right, I'll most probably tackle this once the whole sending of the Check-In is completed (a.k.a after the next PR.)

Reply:
Solution suggestion for the follow; leveraging the active mode reqs to force the ICD to stay in active mode.